### DISCUSSION TITLE
Multi query classes

### DISCUSSION THREAD

**Discussion by joshung** [fastapi/fastapi#9962] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class SearchArgs:
    def __init__(
        self,
        colum: str = Query(),
        expression: str = Query(),
        value: str = Query(),
    ):
        self.colum = colum
        self.expression = expression
        self.value = value


@router.get("/api/v1/search_dataclass", tags=["basic"])
def search(args: list[SearchArgs] = Depends()):
    return {"detail": "search-result", "args": args, "results": {"abc": "def"}}
```

### Description

I want to implement some api like query. I want to query more colum in one time like: _query colum "name" is "eqa" value "James"_ AND _colum "age" is "great_than" value "10"_ AND _more_ AND _more_
I want to make model  like 
`{"colum":"name_of_col",
"expression":"expresion",
"value":"value of colum"
}` 

and in GET api, I want to pass all query param in url.

Any one have implement something like this? Please help me out.
Sorry about my English, Hope you understand.

### Operating System

macOS

### FastAPI Version

0.100

### Python Version

3.10

**Comment by arjwilliams** [comment, marked as answer]
Hi @joshung,

FastAPI doesn't support this natively, as it isn't standard HTTP/REST to be able to group query parameters.

Therefore, you will need to come up with a custom solution. I have provided an example below:

```python
from pydantic import BaseModel, constr, ValidationError
from fastapi import FastAPI, Query, Depends
from fastapi.exceptions import RequestValidationError

app = FastAPI()


class Search(BaseModel):
    column: str
    expression: str
    value: str


class SearchDependency(object):
    _query_pattern = constr(pattern="^[^:]+:[^,]+(?:,[^:]+:[^,]+)*$")

    async def __call__(self, search: list[_query_pattern] = Query(default=())):
        parsed_searches: list[Search] = []

        for search_params in search:
            search_args = {}
            for search_param in search_params.split(","):
                param_expression, param_value = search_param.split(":")
                search_args[param_expression] = param_value

            try:
                parsed_search = Search(**search_args)
            except ValidationError as err:
                raise RequestValidationError(err.errors())

            parsed_searches.append(parsed_search)

        return parsed_searches


@app.get("/api/v1/search_dataclass", tags=["basic"])
def search(args: list[Search] = Depends(SearchDependency())):
    return {"detail": "search-result", "args": args, "results": {"abc": "def"}}
```
```bash
$ curl -i "http://localhost:8000/api/v1/search_dataclass?search=column:name,expression:eqa,value:James&search=column:age,expression:great_than,value:10"
HTTP/1.1 200 OK
...
{"detail":"search-result","args":[{"column":"name","expression":"eqa","value":"James"},{"column":"age","expression":"great_than","value":"10"}],"results":{"abc":"def"}}

$ curl -i "http://localhost:8000/api/v1/search_dataclass?search=column:name"
HTTP/1.1 422 Unprocessable Entity
...
{"detail":[{"type":"missing","loc":["expression"],"msg":"Field required","input":{"column":"name"},"url":"https://errors.pydantic.dev/2.0.3/v/missing"},{"type":"missing","loc":["value"],"msg":"Field required","input":{"column":"name"},"url":"https://errors.pydantic.dev/2.0.3/v/missing"}]}
```

**Reply by joshung** [reply]
thanks for your reply but when I run your code I got error 
` _query_ pattern = constr(pattern="^[^:]+:[^,]+(?:,[^:]+:[^,]+)*$") File "pydantic/types.py", line 443, in pydantic.types.constr TypeError: constr() got an unexpected keyword argument 'pattern'` 
Can you please preview this one for me, I try to upgrade to pydantic 2 but not luck still got problem because sqlmodel req pydantic <2.0.0.0 but it's okie. I will try my best. 

EDIT: I try to use 
`_query_ pattern = re.compile(r"^[^:]+:[^,]+(?:,[^:]+:[^,]+)*$")` 
instead of 
` _query_pattern = constr(pattern="^[^:]+:[^,]+(?:,[^:]+:[^,]+)*$")` 
fixed my erro ✌️

**Reply by arjwilliams** [reply]
Yes, sorry. Pydantic changed the keyword in version 2 from `regex` to `pattern`, and I was assuming you were using version 2.

So instead, you can use:

```python
_query_pattern = constr(regex="^[^:]+:[^,]+(?:,[^:]+:[^,]+)*$")
```

**Reply by joshung** [reply]
thanks to you so much. I solve my problem. Thanks to you again ☺️☺️☺️☺️
